strawberryfields/backends/bosonicbackend/bosoniccircuit.py

```diff
@@ -0,0 +1,833 @@
+        if avg:
+            X = np.diag([np.cos(theta), 1 / np.cos(theta)])
+            Y = np.diag(
+                [
+                    (np.sin(theta) ** 2) * (np.exp(-2 * r_anc)),
+                    (np.tan(theta) ** 2) * (1 - eta_anc) / eta_anc,
+                ]
+            )
+            Y *= self.hbar / 2
+            X2, Y2 = self.expandXY([k], X, Y)
+            self.apply_channel(X2, Y2)
+
+        # Add new ancilla mode, interfere it and measure it
+        # Delete ancilla mode from active list
+        elif not avg:
+            self.add_mode()
+            new_mode = self.nlen - 1
+            self.squeeze(r_anc, 0, new_mode)
+            self.beamsplitter(theta, 0, k, new_mode)
+            self.loss(eta_anc, new_mode)
+            self.phase_shift(np.pi / 2, new_mode)
+            val = self.homodyne(new_mode)
+            self.del_mode(new_mode)
+            self.covs = np.delete(self.covs, [2 * new_mode, 2 * new_mode + 1], axis=1)
+            self.covs = np.delete(self.covs, [2 * new_mode, 2 * new_mode + 1], axis=2)
+            self.means = np.delete(self.means, [2 * new_mode, 2 * new_mode + 1], axis=1)
+            self.nlen = self.nlen - 1
+            self.from_xp = from_xp(self.nlen)
+            self.to_xp = to_xp(self.nlen)
+            self.active = self.active[:new_mode]
+            prefac = -np.tan(theta) / np.sqrt(2 * self.hbar * eta_anc)
+            self.displace(prefac * val[0][0], np.pi / 2, k)
+
+        self.phase_shift(phi / 2, k)
+
+        if not avg:
+            return val
+
+        return None
+
+    def phase_shift(self, phi, k):
+        r"""Implement a phase shift in mode k.
+
+        Args:
+           phi (float): phase
+           k (int): mode to be phase shifted
+        """
+        if self.active[k] is None:
+            raise ValueError("Cannot phase shift mode, mode does not exist")
+
+        rot = symp.expand(symp.rotation(phi), k, self.nlen)
+        self.means = update_means(self.means, rot, self.from_xp)
+        self.covs = update_covs(self.covs, rot, self.from_xp)
+
+    def beamsplitter(self, theta, phi, k, l):
+        r"""Implement a beam splitter operation between modes k and l.
+
+        Args:
+            theta (float): real beamsplitter angle
+            phi (float): complex beamsplitter angle
+            k (int): first mode
+            l (int): second mode
+        """
+        if self.active[k] is None or self.active[l] is None:
+            raise ValueError("Cannot perform beamsplitter, mode(s) do not exist")
+
+        if k == l:
+            raise ValueError("Cannot use the same mode for beamsplitter inputs")
+
+        # Cross-check with Gaussian backend.
+        bs = symp.expand(symp.beam_splitter(theta, phi), [k, l], self.nlen)
+        self.means = update_means(self.means, bs, self.from_xp)
+        self.covs = update_covs(self.covs, bs, self.from_xp)
+
+    def scovmatxp(self):
+        r"""Returns the symmetric ordered array of covariance matrices
```

Comment:
Perhaps worth noting in this and upcoming docstrings what the original ordering was?